flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/architecture/Predicates.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.testutils.architecture;
+
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaMethod;
+import com.tngtech.archunit.core.domain.JavaParameterizedType;
+import com.tngtech.archunit.core.domain.JavaType;
+import com.tngtech.archunit.core.domain.properties.HasName;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** TODO. */
+public class Predicates {
+
+    /** TODO. */
+    public static <T extends HasName> ArchCondition<T> fulfill(DescribedPredicate<T> predicate) {
+        return new ArchCondition<T>(predicate.getDescription()) {
+            @Override
+            public void check(T item, ConditionEvents events) {
+                if (!predicate.apply(item)) {
+                    final String message =
+                            String.format(
+                                    "%s does not satisfy: %s",
+                                    item.getName(), predicate.getDescription());
+                    events.add(SimpleConditionEvent.violated(item, message));
+                }
+            }
+        };
+    }
+
+    /** TODO. */
+    @SafeVarargs
+    public static DescribedPredicate<JavaClass> areAnnotatedWithAtLeastOneOf(
+            Class<? extends Annotation>... annotations) {
+        return new DescribedPredicate<JavaClass>(
+                "be annotated with at least one of "
+                        + Arrays.stream(annotations)
+                                .map(Class::getSimpleName)
+                                .collect(Collectors.joining(", "))) {
+            @Override
+            public boolean apply(JavaClass clazz) {
+                final JavaClass enclosingClazz = getEnclosingClass(clazz).getBaseComponentType();
```

Comment:
This is a bit of a Flink specialty, I guess, since we decorate only the top-level classes, not nested classes directly.